Code/Framework/AzCore/AzCore/Settings/SettingsRegistryOriginTracker.cpp

```diff
@@ -0,0 +1,230 @@
+#include <AzCore/IO/ByteContainerStream.h>
+
+namespace AZ
+{
+    SettingsRegistryOriginTracker::SettingsRegistryOriginTracker(AZ::SettingsRegistryInterface& registry)
+        : m_settingsRegistry(registry)
+    {
+        registry.SetNotifyForMergeOperations(true);
+        m_notifyHandler = m_settingsRegistry.RegisterNotifier(SettingsNotificationHandler(*this));
+    }
+
+    SettingsRegistryOriginTracker::~SettingsRegistryOriginTracker() = default;
+
+    SettingsRegistryOriginTracker::SettingsNotificationHandler::SettingsNotificationHandler(
+        AZ::SettingsRegistryOriginTracker& originTracker)
+        : m_originTracker(originTracker)
+    {
+    }
+
+    void SettingsRegistryOriginTracker::SetTrackingFilter(TrackingFilterCallback filterCallback)
+    {
+        m_trackingFilter = AZStd::move(filterCallback);
+    }
+
+    AZ::SettingsRegistryInterface& SettingsRegistryOriginTracker::GetSettingsRegistry()
+    {
+        return m_settingsRegistry;
+    }
+
+    SettingsRegistryOriginTracker::SettingsNotificationHandler::~SettingsNotificationHandler() = default;
+
+    bool SettingsRegistryOriginTracker::SettingsRegistryOrigin::operator==(const SettingsRegistryOrigin& origin) const
+    {
+        if (m_originFilePath == origin.m_originFilePath && m_settingsKey == origin.m_settingsKey)
+        {
+            if (m_settingsValue.has_value() == origin.m_settingsValue.has_value())
+            {
+                if (m_settingsValue.has_value())
+                {
+                    auto TrimWhitespace = [](char element) -> bool
+                    {
+                        return !::isspace(element);
+                    };
+                    return AZStd::ranges::equal(*m_settingsValue | AZStd::views::filter(TrimWhitespace),
+                        *origin.m_settingsValue | AZStd::views::filter(TrimWhitespace));
+                }
+                else
+                {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    void AZ::SettingsRegistryOriginTracker::SettingsNotificationHandler::operator()(
+        const AZ::SettingsRegistryInterface::NotifyEventArgs& notifyEventArgs)
+    {
+        m_originTracker.AddOrigin(notifyEventArgs.m_jsonKeyPath, notifyEventArgs.m_mergeFilePath);
+    }
+
+    bool SettingsRegistryOriginTracker::FindLastOrigin(AZ::IO::Path& originPath, AZStd::string_view key)
+    {
+        auto populateOriginPath = [&](const AZ::Dom::Path&, const SettingsRegistryOriginStack& stack)
+        {
+            if (!stack.empty())
+            {
+                originPath = stack.front().m_originFilePath;
+                return false;
+            }
+            return true;
+        };
+        AZ::Dom::Path jsonPath = AZ::Dom::Path(key);
+        AZ::Dom::PrefixTreeTraversalFlags traversalFlags = AZ::Dom::PrefixTreeTraversalFlags::ExcludeChildPaths | AZ::Dom::PrefixTreeTraversalFlags::TraverseMostToLeastSpecific;
```

Comment:
nitpick: You can mark this variable as a compile time expression via the `constexpr keyword``
```suggestion
        constexpr AZ::Dom::PrefixTreeTraversalFlags traversalFlags = AZ::Dom::PrefixTreeTraversalFlags::ExcludeChildPaths | AZ::Dom::PrefixTreeTraversalFlags::TraverseMostToLeastSpecific;
```